concurrency/DatanodeDescriptor.java

```diff
@@ -0,0 +1,1052 @@
+    setLastUpdateMonotonic(Time.monotonicNow());
+    rollBlocksScheduled(getLastUpdateMonotonic());
+  }
+
+  private void updateStorageStats(StorageReport[] reports, long cacheCapacity,
+      long cacheUsed, int xceiverCount, int volFailures,
+      VolumeFailureSummary volumeFailureSummary) {
+    long totalCapacity = 0;
+    long totalRemaining = 0;
+    long totalBlockPoolUsed = 0;
+    long totalDfsUsed = 0;
+    long totalNonDfsUsed = 0;
+    Set<DatanodeStorageInfo> failedStorageInfos = null;
+
+    // Decide if we should check for any missing StorageReport and mark it as
+    // failed. There are different scenarios.
+    // 1. When DN is running, a storage failed. Given the current DN
+    //    implementation doesn't add recovered storage back to its storage list
+    //    until DN restart, we can assume volFailures won't decrease
+    //    during the current DN registration session.
+    //    When volumeFailures == this.volumeFailures, it implies there is no
+    //    state change. No need to check for failed storage. This is an
+    //    optimization.  Recent versions of the DataNode report a
+    //    VolumeFailureSummary containing the date/time of the last volume
+    //    failure.  If that's available, then we check that instead for greater
+    //    accuracy.
+    // 2. After DN restarts, volFailures might not increase and it is possible
+    //    we still have new failed storage. For example, admins reduce
+    //    available storages in configuration. Another corner case
+    //    is the failed volumes might change after restart; a) there
+    //    is one good storage A, one restored good storage B, so there is
+    //    one element in storageReports and that is A. b) A failed. c) Before
+    //    DN sends HB to NN to indicate A has failed, DN restarts. d) After DN
+    //    restarts, storageReports has one element which is B.
+    final boolean checkFailedStorages;
+    if (volumeFailureSummary != null && this.volumeFailureSummary != null) {
+      checkFailedStorages = volumeFailureSummary.getLastVolumeFailureDate() >
+          this.volumeFailureSummary.getLastVolumeFailureDate();
+    } else {
+      checkFailedStorages = (volFailures > this.volumeFailures) ||
+          !heartbeatedSinceRegistration;
+    }
+
+    if (checkFailedStorages) {
+      if (this.volumeFailures != volFailures) {
+        LOG.info("Number of failed storages changes from {} to {}",
+            this.volumeFailures, volFailures);
+      }
+      synchronized (storageMap) {
+        failedStorageInfos =
+            new HashSet<>(storageMap.values());
+      }
+    }
+
+    setCacheCapacity(cacheCapacity);
+    setCacheUsed(cacheUsed);
+    setXceiverCount(xceiverCount);
+    this.volumeFailures = volFailures;
+    this.volumeFailureSummary = volumeFailureSummary;
+    for (StorageReport report : reports) {
+
+      DatanodeStorageInfo storage =
+          storageMap.get(report.getStorage().getStorageID());
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 This line may not be thread-safe: **439**. At this line the following field is not protected by synchronization: `storageMap`. Your code typically protects this field with synchronization on object: `storageMap`.

**Fix**  
 Consider protecting the above field with synchronization on object: `storageMap`. For example, like at lines: **278**, **468**, **427**, and **285**.

**More info**  
 The following lines are where the above field is already protected by synchronization on the above object: **278**, **285**, **309**, **427**, **468**, **477**, **491**, **506**, **523**, **964**, **980**, and **995**. This is a call to `get()`. You are already protecting `get()` and `size()`. You are protecting these operations at lines: **468** and **278**.